source/images.py

```diff
@@ -0,0 +1,116 @@
+        shade_min: str = "@#+-."
+    ):
+        """Convert an image file into ASCII art.
+
+        :param file: The image file path, accepts `.jpg`, `.jpeg`, `.png` or any other image format.
+        :param cols: Number of columns in the ascii image, defaults to 80
+        :param scale: defines the adjusted aspect ratio for the ascii image, defaults to 0.43
+        :param resolution: defines resolution, takes in values as percentage of the highest resolution possible with
+               characters from shade_str
+               Special value: 0 (default) is interpreted as mode switch to minimalistic output with characters from
+               shade_min
+        :param shade_str: defaults to a well known pattern of 69 characters used to define the greyscale
+        :param shade_min: defaults to a minimalistic pattern of 5 characters used to define the greyscale
+        """
+        # TODO: error handling when no file is selected
+        self.file = file
+        self.image = PIL.Image.open(file).convert("L")
+        self.cols = cols
+        self.scale = scale
+        self.shade = {
+            "str": shade_str,
+            "str_len": len(shade_str),
+            "min": shade_min,
+            "min_len": len(shade_min),
+            "length": 0,
+            "resolution": max(0, min(resolution, 100))  # Clamp resolution
+        }
+
+    def __repr__(self) -> str:
+        return f'<Image file={self.file} image={self.image}>'
+
+    def __str__(self) -> str:
+        return '\n'.join(self.generate_ascii)
+
+    def g_scale(self, resolution: int = 0) -> str:
+        """
+        Helper function to determine the character set based on the resolution
+
+        :return: a string of characters which corresponds to the luminance in descending order
+        """
+        if resolution == 0:
+            self.shade["length"] = self.shade["min_len"]
+            return self.shade["min"]
+        else:
+            resolution = resolution * self.shade["str_len"] // 100
+            # get evenly spaced indices and return them as a string
+            indices = np.linspace(0, self.shade["str_len"] - 1, resolution, dtype=int)
+            self.shade["length"] = len(indices)
+
+            return ''.join(self.shade["str"][i] for i in indices)
+
+    @staticmethod
+    def get_average(image: PIL.Image) -> float:
+        """Given a PIL Image, return an average value of luminance for the whole image."""
+        im = np.array(image)
+        w, h = im.shape
+        return np.average(im.reshape(w * h))
+
+    @property
+    def generate_ascii(self) -> List[str]:
+        """
+        Convert the opened image into ASCII art
+
+        :return: A list of strings each representing a row of the image.
+        """
+        horizontal, vertical = self.image.size
+        # compute tile dimensions based on aspect ratio and scale
+        w = horizontal / self.cols
+        h = int(w / self.scale)
+        rows = vertical // h
```

Comment:
```suggestion
        w = horizontal // self.cols
        h = w // self.scale
        rows = vertical // h
```

Reply:
tried this earlier, output `rows` is a float for some reason